codegen-v2/src/lib.rs

```diff
@@ -0,0 +1,147 @@
+#[macro_use]
+extern crate serde;
+
+use grammar::{GHeaderFileItem, ParseTree};
+use reader::Reader;
+use std::{
+    collections::HashMap,
+    path::{Path, PathBuf},
+};
+
+mod grammar;
+mod reader;
+#[cfg(test)]
+mod tests;
+
+pub type Result<T> = std::result::Result<T, Error>;
+
+#[derive(Debug)]
+pub enum Error {
+    Todo,
+    Eof,
+}
+
+#[derive(Debug, Clone, Eq, PartialEq, Serialize, Deserialize)]
+pub struct CHeaderDirectory {
+    map: HashMap<PathBuf, Vec<GHeaderFileItem>>,
+}
+
+pub fn parse_dir(path: &Path, mut main_dir: CHeaderDirectory) -> Result<CHeaderDirectory> {
+    let entries = std::fs::read_dir(path).unwrap();
+    for entry in entries {
+        let child_path = entry.unwrap().path();
+
+        if child_path.is_dir() {
+            main_dir = parse_dir(&child_path, main_dir)?;
+        } else if child_path.is_file() {
+            main_dir = parse_file(&child_path, main_dir)?;
+        } else {
+            panic!()
+        }
+    }
+
+    Ok(main_dir)
+}
+
+pub fn parse_file(path: &Path, mut main_dir: CHeaderDirectory) -> Result<CHeaderDirectory> {
+    if !path.is_file() {
+        return Err(Error::Todo);
+    }
+
+    let parent = path.parent().unwrap();
+
+    let content = std::fs::read_to_string(path).map_err(|_err| Error::Todo)?;
+    // TODO: Find a better solution for this.
+    let content: String = content.chars().filter(|c| c.is_ascii()).collect();
+
+    let mut reader = Reader::from(content.as_str());
+    let mut items = vec![];
+
+    loop {
+        let (pending, checked_out) = reader.checkout();
+        if let Ok(der) = GHeaderFileItem::derive(checked_out) {
+            let (derived, branch) = (der.derived, der.branch);
+            reader = pending.merge(branch);
+
+            if let GHeaderFileItem::HeaderInclude(include) = &derived {
+                let child_path = parent.join(Path::new(&include.0));
+
+                // Skip referencing includes.
+                if child_path == path {
+                    continue;
+                }
```

Comment:
I'm just curious how is it possible
Am I right that it should look like
```c++
// foo.h

#include "foo.h"

void foo();
```

Reply:
Yeah, I had to implement this because of this issue: https://github.com/trustwallet/wallet-core/issues/3080

Reply:
Also, the comment should be:

> // Skip **self**-referencing includes.